crates/sui-core/src/checkpoints/causal_order.rs

```diff
@@ -0,0 +1,421 @@
+        // become candidates for sequencing.
+        let mut forward_index: BTreeMap<&TransactionDigest, Vec<&TransactionDigest>> =
+            BTreeMap::new();
+        for (d, effect) in &effect_map {
+            for dep in &effect.dependencies {
+                // We only record the dependencies not in a checkpoint, as the ones
+                // in a checkpoint are already satisfied presumably.
+                if tx_not_in_checkpoint.contains(dep) {
+                    forward_index.entry(dep).or_default().push(d);
+                }
+            }
+        }
+
+        // Define the master sequence, to contain the initial transactions
+        // by transaction digest order.
+        let mut master_sequence: Vec<&TransactionDigest> =
+            initial_transactions.iter().cloned().collect();
+        // A set mirroring the contents of the mater sequence for quick lookup
+        let mut master_set = initial_transactions.clone();
+        // The transactions that just became executed.
+        let mut candidates = initial_transactions;
+
+        // Trace forward the executed transactions, starting from the initial set, and adding more
+        // trasnactions as all their dependencies become executed. The candidates represent executed
+        // transactions that need to have subsequent transactions depending on them examined to
+        // determine if all their dependencies are executed. If so they are sequenced, and also added
+        // to the candidate set to be examiner once.
+        while !candidates.is_empty() {
+            // we continue while we can make progress
+
+            // Take a transaction
+            let next_transaction = *candidates.iter().next().unwrap();
+            candidates.remove(next_transaction);
+
+            // If nothing depends on this tx move on to the next
+            let forward_deps = forward_index.get(next_transaction);
+            if forward_deps.is_none() {
+                continue;
+            }
+
+            // Check all transactions that depend on it, to see if all
+            // dependencies are  satisfied.
+            for dep in forward_deps.unwrap() {
+                // If the forward dependency is not included in the current checkpoint, ignore.
+                if !effect_map.contains_key(*dep) {
+                    continue;
+                }
+
+                // We have already included this into the candidate set once.
+                if master_set.contains(dep) {
+                    continue;
+                }
+
+                if effect_map
+                    .get(*dep)
+                    .unwrap()
+                    .dependencies
+                    .iter()
+                    // All dependencies sequenced in the master seq or in previous checkpoint
+                    .filter(|item| tx_not_in_checkpoint.contains(*item))
+                    .all(|item| master_set.contains(item))
```

Comment:
This is a bit inefficient and not quite intuitive way to do topological sort.. But it's ok to leave it as it is. We can simplify this part latter. Maybe add a TODO

Reply:
Fixed it into a more recognizable topological sort, as mark also suggested.